### PR TITLE
[AArch64][GISel] Translate legal SVE formal arguments and select COPY for SVE

### PR DESCRIPTION
**Pull request by Him188**
This patch adds support for legal SVE fromal arguments in IRTranslator, and support for COPY with SVE.

SVE arguments are allowed only if the hidden option `-aarch64-enable-gisel-sve` is enabled. Illegal types and predicates like `nxv8i1` are not supported yet.

### CODE REVIEW

File: llvm/lib/Target/AArch64/GISel/AArch64RegisterBankInfo.cpp [hunk 1]
```diff
@@ -700,10 +701,10 @@ AArch64RegisterBankInfo::getInstrMapping(const MachineInstr &MI) const {
       // If both RB are null that means both registers are generic.
       // We shouldn't be here.
       assert(DstRB && SrcRB && "Both RegBank were nullptr");
-      unsigned Size = getSizeInBits(DstReg, MRI, TRI);
+      TypeSize Size = getSizeInBits(DstReg, MRI, TRI);
       return getInstructionMapping(
-          DefaultMappingID, copyCost(*DstRB, *SrcRB, TypeSize::getFixed(Size)),
-          getCopyMapping(DstRB->getID(), SrcRB->getID(), Size),
+          DefaultMappingID, copyCost(*DstRB, *SrcRB, Size),
+          getCopyMapping(DstRB->getID(), SrcRB->getID(), Size.getKnownMinValue()),
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/GISel/AArch64RegisterBankInfo.cpp:707]: Could we change getCopyMapping to take a TypeSize too?

File: llvm/lib/Target/AArch64/GISel/AArch64CallLowering.cpp [hunk 2]
```diff
@@ -53,6 +53,8 @@
 using namespace llvm;
 using namespace AArch64GISelUtils;
 
+extern cl::opt<bool> EnableSVEGISel;
```
Comment (arsenm) [inline llvm/lib/Target/AArch64/GISel/AArch64CallLowering.cpp:56]: This shouldn't need to spread. I don't like having fallBackToDAGISel at all, but I see we have 2 of them. Why not just put them both in the same place?

File: llvm/test/CodeGen/AArch64/GlobalISel/translate-sve-formal-argument.ll [hunk 3]
```diff
@@ -0,0 +1,389 @@
+; NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=aarch64-linux-gnu -O0 -mattr=+sve -global-isel -global-isel-abort=1 -aarch64-enable-gisel-sve=1 \
+; RUN:     -stop-after=irtranslator -verify-machineinstrs %s -o - | FileCheck %s
+
+;; vscale x 128-bit
```
Comment (tschuett) [inline llvm/test/CodeGen/AArch64/GlobalISel/translate-sve-formal-argument.ll:5]: Could you add tests with several legal arguments to show the correct usage of the z registers?

Comment (Him188) [reply llvm/test/CodeGen/AArch64/GlobalISel/translate-sve-formal-argument.ll:5]: Isn't [sve-formal-argument.ll](https://github.com/llvm/llvm-project/pull/95236/files#diff-e38af48ccacbc706e49c732d40e671d7b38f6296861cb2947728bd95e8b56d8c) doing this?

Comment (tschuett) [reply llvm/test/CodeGen/AArch64/GlobalISel/translate-sve-formal-argument.ll:5]: Yeah. Maybe. But it is assembler.

File: llvm/lib/Target/AArch64/GISel/AArch64CallLowering.cpp (hunk 2 above)
Comment (Him188) [reply llvm/lib/Target/AArch64/GISel/AArch64CallLowering.cpp:56]: `AArch64CallLowering::fallBackToDAGISel` works on `MachineFunction` as an early path before IRTranslate. 
`AArch64TargetLowering::fallBackToDAGISel` works on each `Instruction`.

FastISel also uses `AArch64TargetLowering::fallBackToDAGISel`, making it more complex to merge the logic.

This patch does not completely support SVE formal arguments, especially for predicate registers <vscale x 16 x i1>. 
The option is needed until we support them.

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp [hunk 4]
```diff
@@ -597,8 +597,14 @@ getRegClassForTypeOnBank(LLT Ty, const RegisterBank &RB,
 /// Given a register bank, and size in bits, return the smallest register class
 /// that can represent that combination.
 static const TargetRegisterClass *
-getMinClassForRegBank(const RegisterBank &RB, unsigned SizeInBits,
+getMinClassForRegBank(const RegisterBank &RB, TypeSize SizeInBits,
```
Comment (madhur13490) [inline llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:600]: The variable name "SizeInBits" is not appropriate for Type "TypeSize". Please have something else.

File: llvm/test/CodeGen/AArch64/GlobalISel/sve-formal-argument.ll [hunk 5]
```diff
@@ -0,0 +1,45 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=aarch64-linux-gnu -O0 -mattr=+sve -global-isel -global-isel-abort=1 -aarch64-enable-gisel-sve=1 %s -o - | FileCheck %s
```
Comment (madhur13490) [inline llvm/test/CodeGen/AArch64/GlobalISel/sve-formal-argument.ll:2]: Why with -O0?

**Review by madhur13490 — COMMENTED** [review]
Does this pass all regression tests? I had the same patch locally but I was seeing some failures. May be my patch had some issues...

**Comment by Him188** [comment]
> Does this pass all regression tests? 

@madhur13490  Yes, `check-all` passed. The failing CI is due to failed LoongArch test.

File: llvm/lib/Target/AArch64/GISel/AArch64CallLowering.cpp [hunk 6]
```diff
@@ -525,10 +527,10 @@ static void handleMustTailForwardedRegisters(MachineIRBuilder &MIRBuilder,
 
 bool AArch64CallLowering::fallBackToDAGISel(const MachineFunction &MF) const {
   auto &F = MF.getFunction();
-  if (F.getReturnType()->isScalableTy() ||
+  if (!EnableSVEGISel && (F.getReturnType()->isScalableTy() ||
```
Comment (tschuett) [inline llvm/lib/Target/AArch64/GISel/AArch64CallLowering.cpp:530]: Are scalable return types enabled now?

Comment (Him188) [reply llvm/lib/Target/AArch64/GISel/AArch64CallLowering.cpp:530]: No. I'm following the same idea from https://github.com/llvm/llvm-project/pull/92130#discussion_r1600378977 to enable SVE only with a debug option

File: llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp (hunk 4 above)
Comment (Him188) [reply llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp:600]: I think it's fine, as it's obtained from `TypeSize RegisterBankInfo::getSizeInBits()`. 

There are both `TypeSize LLT::getSizeInBytes()` and `TypeSize LLT::getSizeInBits()` so `TypeSize` itself does not have a unit.

File: llvm/lib/Target/AArch64/GISel/AArch64RegisterBankInfo.cpp (hunk 1 above)
Comment (Him188) [reply llvm/lib/Target/AArch64/GISel/AArch64RegisterBankInfo.cpp]: Sounds better. Done

File: llvm/test/CodeGen/AArch64/GlobalISel/translate-sve-formal-argument.ll (hunk 3 above)
Comment (Him188) [reply llvm/test/CodeGen/AArch64/GlobalISel/translate-sve-formal-argument.ll:5]: Added more tests

**Review by davemgreen — APPROVED** [review]
I'm not 100% sure about all the RegBankInfo stuff, but we can change that if we need to later. From what I can tell this LGTM if the tests pass with a rebase.

**Comment by davemgreen** [comment]
The CI seems to be a bit flakey lately. Feel free to commit if the failures look unrelated and nothing fails locally.

**Pull request merged**